src/frequenz/sdk/timeseries/_resampling.py

```diff
@@ -483,6 +498,10 @@ async def resample(self, *, one_shot: bool = False) -> None:
         async for drift in self._timer:
             now = datetime.now(tz=timezone.utc)
 
+            # resync the window_end to the system time
+            if (self._window_end - now) > self._config.resampling_period:
+                self._window_end = self._sync_timer()
```

Comment:
> And this looks a bit weird to me, so under normal circumstances (no wall clock change), you want to resync the window_end?

Not quite. We want to update the next tick time in the timer and adjust the time window that is used for resampling.
The `self._window_end` can be seen as the resamplers internal time, which (before the patch) was only monotonic increased.

This check is added to calculate the drift of the internal clock, i.e. `self.window_end` against the wall clock.

> In the case of the wall clock changing forward (so if window_end, we is 10 and now should have been 10 but it's 12 because the wall clock jumped 2 seconds forward), then drift_relative_to_wall_clock is 2, assuming a resampling_period, p, of 1, then we won't be synced?
> 
> And in the case of the wall clock changing backwards (so if window_end, we is 10 and now should have been 10 but it's 8 because the wall clock jumped 2 seconds backwards), then drift_relative_to_wall_clock is -2, assuming a resampling_period, p, of 1, then we will be synced?

I think it would be the other way around, but in any case good catch. It should resync as soon as the window_end drifts away more then one sample period from the wall clock, since that means that at least one sample drifted out of the resampling window.
I remember I had an `abs(...)` in there. Maybe it got lost while rebasing (I had a lot more code for testing etc. that got reduced to these two lines).

I'll update the variable names and add the abs back.

Reply:
I see a problem with that patch. In fact, whenever we resync the window_end is new calculated to a position that is a bit in the future, such that the resampler gets some time to collect some samples.

If we now resync because of an `async drift` the resampler would leave out one value.

I see two solutions.

1) Don't re-sync when drift is above the tolerance.
2) Don't add these extra delay.
3) Have different sync codes for initialization and for re-sync

I prefer option 2) because I don't understand why we do this. If the resampler starts as soon as possible after initialization and there are no samples then it sends out nones.